File: web/src/main/java/hu/bsstudio/robonaut/speed/UpdateSpeedTimerHandler.java
package hu.bsstudio.robonaut.speed;

import hu.bsstudio.robonaut.race.speed.timer.SpeedTimerService;
import hu.bsstudio.robonaut.race.speed.timer.model.SpeedTimer;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.springframework.web.reactive.function.server.HandlerFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
public class UpdateSpeedTimerHandler implements HandlerFunction<ServerResponse> {

    @NonNull
    private final SpeedTimerService service;

    @Override
    public Mono<ServerResponse> handle(final ServerRequest request) {
        final var speedTimer = request.bodyToMono(SpeedTimer.class)
            .flatMap(service::updateTimer);
        return ServerResponse.ok().body(speedTimer, SpeedTimer.class);
    }
}

File: web/src/main/kotlin/hu/bsstudio/robonaut/speed/UpdateSpeedTimerHandler.kt
package hu.bsstudio.robonaut.speed

import hu.bsstudio.robonaut.race.speed.timer.SpeedTimerService
import hu.bsstudio.robonaut.race.speed.timer.model.SpeedTimer
import org.springframework.web.reactive.function.server.*
import reactor.core.publisher.Mono

class UpdateSpeedTimerHandler(private val service: SpeedTimerService) : HandlerFunction<ServerResponse> {

    override fun handle(request: ServerRequest): Mono<ServerResponse> {
        return request.bodyToMono<SpeedTimer>()
            .flatMap(service::updateTimer)
            .let { speedTimer -> ServerResponse.ok().body(speedTimer) }
    }
}
